In [3]:
import pandas as pd
import json
import numpy as np
from matplotlib import pyplot as plt
from scipy import stats
import base64
import requests

In [4]:
keys = pd.read_csv("Spotify-Keys.txt", header = 0, sep = ",")
appid = keys['Client_ID'][0]
appsecret = keys['Client_Secret'][0]
client_cred = base64.b64encode(str(appid + ":" + appsecret).encode('ascii'))
header = {'Authorization': 'Basic {}'.format(client_cred.decode('ascii'))}

payload = {'grant_type' : 'client_credentials'}
access_token_url = "https://accounts.spotify.com/api/token"
response = requests.post(access_token_url, headers = header, data = payload)
session_token = response.json()['access_token']
session_header = {'Authorization': 'Bearer {}'.format(response.json()['access_token'])}

### Album 1: Put Up Or Shut Up

In [5]:
PutUp_id = '47boQgN4dboq2f0IcvI4Bt'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(PutUp_id)
PutUp_tracks = requests.get(album_tracks_endpoint, headers = session_header)
PutUp_dict = json.loads(PutUp_tracks.text)
PutUp_track_ids = [x['id'] for x in PutUp_dict['items']]
PutUp_track_ids
PutUp_track_strings = ','.join(str(x) for x in PutUp_track_ids)
PutUp_track_strings
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(PutUp_track_strings)
PutUp_features = requests.get(track_features_endpoint, headers = session_header)
PutUp_dict = json.loads(PutUp_features.text)

In [6]:
PutUp_dict['audio_features'][0]

{'danceability': 0.425,
 'energy': 0.852,
 'key': 9,
 'loudness': -6.148,
 'mode': 1,
 'speechiness': 0.122,
 'acousticness': 0.00228,
 'instrumentalness': 0,
 'liveness': 0.0963,
 'valence': 0.431,
 'tempo': 185.021,
 'type': 'audio_features',
 'id': '3j1HTRodSjGtK3YsYMI0XF',
 'uri': 'spotify:track:3j1HTRodSjGtK3YsYMI0XF',
 'track_href': 'https://api.spotify.com/v1/tracks/3j1HTRodSjGtK3YsYMI0XF',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/3j1HTRodSjGtK3YsYMI0XF',
 'duration_ms': 181027,
 'time_signature': 4}

In [7]:
PutUp_df = pd.DataFrame(PutUp_dict['audio_features'])

In [8]:
PutUp_df['album'] = ['PutUp'] * len(PutUp_df.index)

In [9]:
PutUp_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.425   0.852    9    -6.148     1       0.1220      0.002280   
1         0.352   0.862    4    -6.521     1       0.0585      0.000042   
2         0.253   0.825    7    -6.903     1       0.0573      0.009480   
3         0.486   0.901    2    -6.943     1       0.0394      0.000049   
4         0.401   0.945    4    -5.984     1       0.2160      0.001880   
5         0.458   0.796    9    -6.993     1       0.0597      0.000959   
6         0.286   0.894    4    -6.393     1       0.0950      0.001320   

   instrumentalness  liveness  valence    tempo            type  \
0                 0    0.0963    0.431  185.021  audio_features   
1                 0    0.2980    0.491  179.769  audio_features   
2                 0    0.1140    0.115  172.679  audio_features   
3                 0    0.0828    0.353   96.030  audio_features   
4                 0    0.3250    0.408  174.953  audio_features   
5                 0    0.3010    0.520  176.944  audio_features   
6                 0    0.1070    0.132  104.335  audio_features   

                       id                                   uri  \
0  3j1HTRodSjGtK3YsYMI0XF  spotify:track:3j1HTRodSjGtK3YsYMI0XF   
1  6b23XfrLswilUCsHQHPY1i  spotify:track:6b23XfrLswilUCsHQHPY1i   
2  5ZEQLXv9F2pKtQrGwMKpMB  spotify:track:5ZEQLXv9F2pKtQrGwMKpMB   
3  1oTtOTbcP8waTv4wPsCqSm  spotify:track:1oTtOTbcP8waTv4wPsCqSm   
4  6f7h261u7CPBNFRBYL5Boq  spotify:track:6f7h261u7CPBNFRBYL5Boq   
5  6PTr9xi7S1eZskFIjPNY8e  spotify:track:6PTr9xi7S1eZskFIjPNY8e   
6  0SQheC3TXkqJY6zZbWMv9v  spotify:track:0SQheC3TXkqJY6zZbWMv9v   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/3j1HTRodSjGt...   
1  https://api.spotify.com/v1/tracks/6b23XfrLswil...   
2  https://api.spotify.com/v1/tracks/5ZEQLXv9F2pK...   
3  https://api.spotify.com/v1/tracks/1oTtOTbcP8wa...   
4  https://api.spotify.com/v1/tracks/6f7h261u7CPB...   
5  https://api.spotify.com/v1/tracks/6PTr9xi7S1eZ...   
6  https://api.spotify.com/v1/tracks/0SQheC3TXkqJ...   

                                        analysis_url  duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/3j1H...       181027   
1  https://api.spotify.com/v1/audio-analysis/6b23...       183573   
2  https://api.spotify.com/v1/audio-analysis/5ZEQ...       218200   
3  https://api.spotify.com/v1/audio-analysis/1oTt...       218293   
4  https://api.spotify.com/v1/audio-analysis/6f7h...       177840   
5  https://api.spotify.com/v1/audio-analysis/6PTr...       180693   
6  https://api.spotify.com/v1/audio-analysis/0SQh...       242973   

   time_signature  album  
0               4  PutUp  
1               4  PutUp  
2               4  PutUp  
3               4  PutUp  
4               4  PutUp  
5               4  PutUp  
6               3  PutUp

## Album 2: So Wrong It's Right

In [10]:
SoWrongItsRight_id = '0H4ELcHequ3OligrjWfrZP'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(SoWrongItsRight_id)
SoWrongItsRight_tracks = requests.get(album_tracks_endpoint, headers = session_header)
SoWrongItsRight_dict = json.loads(SoWrongItsRight_tracks.text)
SoWrongItsRight_track_ids = [x['id'] for x in SoWrongItsRight_dict['items']]
SoWrongItsRight_track_ids
SoWrongItsRight_track_strings = ','.join(str(x) for x in SoWrongItsRight_track_ids)
SoWrongItsRight_track_strings

'2rQSAtxX1yuoWq8ZrvUhpL,4NpzB4z4B0UcajopWE5e6h,3MfRql0HwuxEZzGpoIC832,62wWvHpBOmbGbzkls0Iomo,3xOE5ciDmfQ0VY3WahE0eJ,0JJP0IS4w0fJx01EcrfkDe,5pWqD58cZ9TNvL9WbNfcQ3,2qxXypNXOJZ5qUFdpzJ56n,64WvIOjnquE5dasQ7Kb7Bs,1Nhb2W4BXDTdSbNBw0BxVd,3RtBhVKXnctRgS5Bz61svC,3aCRrvADMjFjZq1x8dWD0t'

In [11]:
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(SoWrongItsRight_track_strings)

In [12]:
SoWrongItsRight_features = requests.get(track_features_endpoint, headers = session_header)

In [13]:
SoWrongItsRight_dict = json.loads(SoWrongItsRight_features.text)

In [14]:
SoWrongItsRight_dict['audio_features'][0]

{'danceability': 0.312,
 'energy': 0.953,
 'key': 4,
 'loudness': -2.051,
 'mode': 1,
 'speechiness': 0.13,
 'acousticness': 0.0035,
 'instrumentalness': 0,
 'liveness': 0.348,
 'valence': 0.644,
 'tempo': 198.898,
 'type': 'audio_features',
 'id': '2rQSAtxX1yuoWq8ZrvUhpL',
 'uri': 'spotify:track:2rQSAtxX1yuoWq8ZrvUhpL',
 'track_href': 'https://api.spotify.com/v1/tracks/2rQSAtxX1yuoWq8ZrvUhpL',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/2rQSAtxX1yuoWq8ZrvUhpL',
 'duration_ms': 149613,
 'time_signature': 4}

In [15]:
SoWrongItsRight_df = pd.DataFrame(SoWrongItsRight_dict['audio_features'])

In [16]:
SoWrongItsRight_df['album'] = ['SoWrongItsRight'] * len(SoWrongItsRight_df.index)

In [17]:
SoWrongItsRight_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.312   0.953    4    -2.051     1       0.1300      0.003500   
1          0.373   0.952    9    -2.525     1       0.0813      0.002010   
2          0.488   0.967    3    -2.701     1       0.0755      0.014700   
3          0.343   0.944   11    -3.124     1       0.0559      0.000853   
4          0.404   0.968    0    -2.211     1       0.1190      0.002300   
5          0.459   0.895    2    -3.126     1       0.0805      0.007250   
6          0.318   0.936    2    -2.522     1       0.0489      0.001710   
7          0.450   0.602    6    -4.801     0       0.0264      0.150000   
8          0.456   0.899    7    -3.799     1       0.0636      0.002680   
9          0.318   0.908    3    -2.286     1       0.0623      0.011900   
10         0.423   0.945    4    -2.572     1       0.1180      0.002710   
11         0.582   0.880    2    -3.219     1       0.0601      0.016500   

    instrumentalness  liveness  valence    tempo            type  \
0                  0    0.3480    0.644  198.898  audio_features   
1                  0    0.0355    0.592  182.963  audio_features   
2                  0    0.1090    0.341  130.928  audio_features   
3                  0    0.0896    0.482  179.891  audio_features   
4                  0    0.1280    0.639  167.012  audio_features   
5                  0    0.2060    0.572  181.040  audio_features   
6                  0    0.0852    0.587  182.996  audio_features   
7                  0    0.1410    0.316  170.043  audio_features   
8                  0    0.2150    0.528  175.054  audio_features   
9                  0    0.0705    0.459  163.973  audio_features   
10                 0    0.3070    0.561  164.982  audio_features   
11                 0    0.3890    0.521  123.013  audio_features   

                        id                                   uri  \
0   2rQSAtxX1yuoWq8ZrvUhpL  spotify:track:2rQSAtxX1yuoWq8ZrvUhpL   
1   4NpzB4z4B0UcajopWE5e6h  spotify:track:4NpzB4z4B0UcajopWE5e6h   
2   3MfRql0HwuxEZzGpoIC832  spotify:track:3MfRql0HwuxEZzGpoIC832   
3   62wWvHpBOmbGbzkls0Iomo  spotify:track:62wWvHpBOmbGbzkls0Iomo   
4   3xOE5ciDmfQ0VY3WahE0eJ  spotify:track:3xOE5ciDmfQ0VY3WahE0eJ   
5   0JJP0IS4w0fJx01EcrfkDe  spotify:track:0JJP0IS4w0fJx01EcrfkDe   
6   5pWqD58cZ9TNvL9WbNfcQ3  spotify:track:5pWqD58cZ9TNvL9WbNfcQ3   
7   2qxXypNXOJZ5qUFdpzJ56n  spotify:track:2qxXypNXOJZ5qUFdpzJ56n   
8   64WvIOjnquE5dasQ7Kb7Bs  spotify:track:64WvIOjnquE5dasQ7Kb7Bs   
9   1Nhb2W4BXDTdSbNBw0BxVd  spotify:track:1Nhb2W4BXDTdSbNBw0BxVd   
10  3RtBhVKXnctRgS5Bz61svC  spotify:track:3RtBhVKXnctRgS5Bz61svC   
11  3aCRrvADMjFjZq1x8dWD0t  spotify:track:3aCRrvADMjFjZq1x8dWD0t   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/2rQSAtxX1yuo...   
1   https://api.spotify.com/v1/tracks/4NpzB4z4B0Uc...   
2   https://api.spotify.com/v1/tracks/3MfRql0HwuxE...   
3   https://api.spotify.com/v1/tracks/62wWvHpBOmbG...   
4   https://api.spotify.com/v1/tracks/3xOE5ciDmfQ0...   
5   https://api.spotify.com/v1/tracks/0JJP0IS4w0fJ...   
6   https://api.spotify.com/v1/tracks/5pWqD58cZ9TN...   
7   https://api.spotify.com/v1/tracks/2qxXypNXOJZ5...   
8   https://api.spotify.com/v1/tracks/64WvIOjnquE5...   
9   https://api.spotify.com/v1/tracks/1Nhb2W4BXDTd...   
10  https://api.spotify.com/v1/tracks/3RtBhVKXnctR...   
11  https://api.spotify.com/v1/tracks/3aCRrvADMjFj...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/2rQS...       149613   
1   https://api.spotify.com/v1/audio-analysis/4Npz...       180440   
2   https://api.spotify.com/v1/audio-analysis/3MfR...       216547   
3   https://api.spotify.com/v1/audio-analysis/62wW...       231973   
4   https://api.spotify.com/v1/audio-analysis/3xOE...       181720   
5   https://api.spotify.com/v1/audio-analysis/0JJP...       182827   
6   https://a

### Album 3: Nothing Personal (Deluxe Version)

In [18]:
NothingPersonal_id = '7tY7NcC2gKTxXsTJM6zK5I'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(NothingPersonal_id)
NothingPersonal_tracks = requests.get(album_tracks_endpoint, headers = session_header)
NothingPersonal_dict = json.loads(NothingPersonal_tracks.text)
NothingPersonal_track_ids = [x['id'] for x in NothingPersonal_dict['items']]
NothingPersonal_track_ids
NothingPersonal_track_strings = ','.join(str(x) for x in NothingPersonal_track_ids)
NothingPersonal_track_strings
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(NothingPersonal_track_strings)
NothingPersonal_features = requests.get(track_features_endpoint, headers = session_header)
NothingPersonal_dict = json.loads(NothingPersonal_features.text)

In [19]:
NothingPersonal_dict['audio_features'][0]

{'danceability': 0.497,
 'energy': 0.923,
 'key': 2,
 'loudness': -4.512,
 'mode': 1,
 'speechiness': 0.138,
 'acousticness': 0.00282,
 'instrumentalness': 0,
 'liveness': 0.158,
 'valence': 0.663,
 'tempo': 181.074,
 'type': 'audio_features',
 'id': '0aZJGkkXR3DgaFqo5sB8ot',
 'uri': 'spotify:track:0aZJGkkXR3DgaFqo5sB8ot',
 'track_href': 'https://api.spotify.com/v1/tracks/0aZJGkkXR3DgaFqo5sB8ot',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/0aZJGkkXR3DgaFqo5sB8ot',
 'duration_ms': 198000,
 'time_signature': 4}

In [20]:
NothingPersonal_df = pd.DataFrame(NothingPersonal_dict['audio_features'])

In [21]:
NothingPersonal_df['album'] = ['NothingPersonal'] * len(NothingPersonal_df.index)

In [22]:
NothingPersonal_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.497   0.923    2    -4.512     1       0.1380      0.002820   
1          0.417   0.951    1    -3.006     1       0.0499      0.002950   
2          0.554   0.902    4    -4.019     1       0.0566      0.006360   
3          0.473   0.963   10    -3.795     1       0.0571      0.003090   
4          0.498   0.945    5    -3.555     1       0.0372      0.025000   
5          0.483   0.776   11    -4.416     0       0.0334      0.002390   
6          0.496   0.893    9    -3.201     1       0.0705      0.072500   
7          0.489   0.967    7    -3.168     1       0.1510      0.007950   
8          0.580   0.812    8    -4.868     1       0.0429      0.046100   
9          0.396   0.985    4    -2.800     1       0.1280      0.003300   
10         0.548   0.958    3    -2.981     1       0.0411      0.014000   
11         0.313   0.713    9    -4.489     1       0.0327      0.025800   
12         0.406   0.972    9    -3.533     0       0.3390      0.000756   
13         0.573   0.806    3    -5.020     1       0.0432      0.000035   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000000    0.1580    0.663  181.074  audio_features   
1           0.000000    0.3350    0.643  147.944  audio_features   
2           0.000000    0.3280    0.662  126.021  audio_features   
3           0.000000    0.2440    0.582  150.033  audio_features   
4           0.000000    0.2780    0.559   81.508  audio_features   
5           0.000000    0.3180    0.378   96.953  audio_features   
6           0.000000    0.0644    0.731  148.015  audio_features   
7           0.000000    0.0616    0.397  153.034  audio_features   
8           0.000000    0.0968    0.364  120.012  audio_features   
9           0.000000    0.4810    0.517  166.016  audio_features   
10          0.000000    0.1220    0.608   98.982  audio_features   
11          0.000000    0.1350    0.196  169.871  audio_features   
12          0.000000    0.0658    0.199  164.978  audio_features   
13          0.000284    0.2700    0.549  150.078  audio_features   

                        id                                   uri  \
0   0aZJGkkXR3DgaFqo5sB8ot  spotify:track:0aZJGkkXR3DgaFqo5sB8ot   
1   3d3rJ0qtOU5aNX324mNWfo  spotify:track:3d3rJ0qtOU5aNX324mNWfo   
2   4CDlYQu16nUtVGwgwqj3I3  spotify:track:4CDlYQu16nUtVGwgwqj3I3   
3   2WO1QXoR1o0R1rYKAxF7Ya  spotify:track:2WO1QXoR1o0R1rYKAxF7Ya   
4   6hmEzWnA6ny2rqVOpB5Hcb  spotify:track:6hmEzWnA6ny2rqVOpB5Hcb   
5   0iQ8rIFYUUJtCV20smTEHR  spotify:track:0iQ8rIFYUUJtCV20smTEHR   
6   3BuArnDZqzmnSr7S0CpQfb  spotify:track:3BuArnDZqzmnSr7S0CpQfb   
7   35CC1Zlq9bvr3TkZtnOAv7  spotify:track:35CC1Zlq9bvr3TkZtnOAv7   
8   5CKVI4DEqIpfebT6kIWHf3  spotify:track:5CKVI4DEqIpfebT6kIWHf3   
9   035SyYaZBffdxtuHRPE5SD  spotify:track:035SyYaZBffdxtuHRPE5SD   
10  1RhrMPd8wdpM2gTI6aiUQA  spotify:track:1RhrMPd8wdpM2gTI6aiUQA   
11  371Tb0eNfpjpSqggjcLv8c  spotify:track:371Tb0eNfpjpSqggjcLv8c   
12  4oCaVJ2DKPjUfOwSUihGEj  spotify:track:4oCaVJ2DKPjUfOwSUihGEj   
13  3sgbLZvnhNK6ZerZqKE9D7  spotify:track:3sgbLZvnhNK6ZerZqKE9D7   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/0aZJGkkXR3Dg...   
1   https://api.spotify.com/v1/tracks/3d3rJ0qtOU5a...   
2   https://api.spotify.com/v1/tracks/4CDlYQu16nUt...   
3   https://api.spotify.com/v1/tracks/2WO1QXoR1o0R...   
4   https://api.spotify.com/v1/tracks/6hmEzWnA6ny2...   
5   https://api.spotify.com/v1/tracks/0iQ8rIFYUUJt...   
6   https://api.spotify.com/v1/tracks/3BuArnDZqzmn...   
7   https://api.spotify.com/v1/tracks/35CC1Zlq9bvr...   
8   https://api.spotify.com/v1/tracks/5CKVI4DEqIpf...   
9   https://api.spotify.com/v1/tracks/035SyYaZBffd...   
10  https://api.spotify.com/v1/tracks/1RhrMPd8wdpM...   
11  https://api.spotify.com/v1/tracks/371Tb0eNfpjp...   
12  https://api.spotify.com/v1/tracks/4oCaVJ2DKPjU...   
13  https://api.spotify

### Album 4: Dirty Work

In [23]:
DirtyWork_id = '2LJKiBomrpmLQhCZmNLrP8'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(DirtyWork_id)
DirtyWork_tracks = requests.get(album_tracks_endpoint, headers = session_header)
DirtyWork_dict = json.loads(DirtyWork_tracks.text)
DirtyWork_track_ids = [x['id'] for x in DirtyWork_dict['items']]
DirtyWork_track_ids
DirtyWork_track_strings = ','.join(str(x) for x in DirtyWork_track_ids)
DirtyWork_track_strings
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(DirtyWork_track_strings)
DirtyWork_features = requests.get(track_features_endpoint, headers = session_header)
DirtyWork_dict = json.loads(DirtyWork_features.text)

In [24]:
DirtyWork_dict['audio_features'][0]

{'danceability': 0.555,
 'energy': 0.98,
 'key': 5,
 'loudness': -2.384,
 'mode': 1,
 'speechiness': 0.0619,
 'acousticness': 0.000871,
 'instrumentalness': 0,
 'liveness': 0.264,
 'valence': 0.529,
 'tempo': 133.066,
 'type': 'audio_features',
 'id': '0jIctjcQRcF1UBDWtJi290',
 'uri': 'spotify:track:0jIctjcQRcF1UBDWtJi290',
 'track_href': 'https://api.spotify.com/v1/tracks/0jIctjcQRcF1UBDWtJi290',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/0jIctjcQRcF1UBDWtJi290',
 'duration_ms': 164947,
 'time_signature': 4}

In [25]:
DirtyWork_df = pd.DataFrame(DirtyWork_dict['audio_features'])

In [26]:
DirtyWork_df['album'] = ['DirtyWork'] * len(DirtyWork_df.index)

In [27]:
DirtyWork_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.555   0.980    5    -2.384     1       0.0619      0.000871   
1          0.512   0.947    1    -2.730     0       0.0817      0.006330   
2          0.624   0.928    4    -3.022     1       0.0575      0.011300   
3          0.411   0.970   11    -2.860     0       0.1070      0.004550   
4          0.520   0.911   11    -3.945     0       0.0550      0.003880   
5          0.578   0.954    2    -3.668     1       0.1170      0.030000   
6          0.477   0.991    4    -2.433     1       0.0923      0.001830   
7          0.406   0.947    9    -3.307     0       0.0944      0.017900   
8          0.502   0.815    1    -4.514     1       0.0388      0.007050   
9          0.536   0.713    2    -4.348     1       0.0294      0.060000   
10         0.545   0.946   10    -3.156     1       0.0387      0.012300   
11         0.541   0.959    0    -3.291     1       0.1020      0.014100   
12         0.418   0.965    5    -2.959     0       0.0956      0.060500   
13         0.588   0.890    7    -3.336     1       0.0411      0.020100   
14         0.574   0.965    4    -3.036     1       0.0692      0.013800   
15         0.597   0.950   11    -2.193     1       0.0521      0.013800   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000000    0.2640    0.529  133.066  audio_features   
1           0.000000    0.0613    0.553  184.005  audio_features   
2           0.000000    0.1020    0.671  132.045  audio_features   
3           0.000000    0.1290    0.488  180.023  audio_features   
4           0.000002    0.3400    0.614  139.046  audio_features   
5           0.000000    0.3250    0.587   91.037  audio_features   
6           0.000000    0.3040    0.560  152.940  audio_features   
7           0.000000    0.1060    0.269  160.769  audio_features   
8           0.000000    0.1960    0.475  111.005  audio_features   
9           0.000003    0.1010    0.547  169.864  audio_features   
10          0.000000    0.3120    0.908   91.970  audio_features   
11          0.000000    0.1420    0.637  140.982  audio_features   
12          0.000000    0.0261    0.760  178.038  audio_features   
13          0.000000    0.0810    0.657  106.038  audio_features   
14          0.000000    0.4300    0.781  128.034  audio_features   
15          0.000000    0.0415    0.898   93.488  audio_features   

                        id                                   uri  \
0   0jIctjcQRcF1UBDWtJi290  spotify:track:0jIctjcQRcF1UBDWtJi290   
1   3PMcBzxNMaW4cVX7FKXwwA  spotify:track:3PMcBzxNMaW4cVX7FKXwwA   
2   3LFzVLgW5N9RYirdQOOPX8  spotify:track:3LFzVLgW5N9RYirdQOOPX8   
3   7xapl7fsqoJU4p0mQVOSm0  spotify:track:7xapl7fsqoJU4p0mQVOSm0   
4   2qqn52sp9dDk2F9PdTrpwZ  spotify:track:2qqn52sp9dDk2F9PdTrpwZ   
5   00SsZ9KAuNQBUYfqRbxBn4  spotify:track:00SsZ9KAuNQBUYfqRbxBn4   
6   2LhR8HU0lswjen4LYVMpsZ  spotify:track:2LhR8HU0lswjen4LYVMpsZ   
7   5X5r8ZUSMaLFfm5zQpaY1R  spotify:track:5X5r8ZUSMaLFfm5zQpaY1R   
8   7Aoyeqo8DFDN8sXC51jQbH  spotify:track:7Aoyeqo8DFDN8sXC51jQbH   
9   3ShzFbJSRNntv62XqTW50Q  spotify:track:3ShzFbJSRNntv62XqTW50Q   
10  4BbagSITXLM4o9u2ImJzKt  spotify:track:4BbagSITXLM4o9u2ImJzKt   
11  5Z3Ym26dWTp63248evI9zM  spotify:track:5Z3Ym26dWTp63248evI9zM   
12  3tOMfrPR21mgkR1KuuVIh7  spotify:track:3tOMfrPR21mgkR1KuuVIh7   
13  7zYgTT9gsTAZtaOpg2IYRQ  spotify:track:7zYgTT9gsTAZtaOpg2IYRQ   
14  1Y2GZOHagYKkKRzNVbpbT5  spotify:track:1Y2GZOHagYKkKRzNVbpbT5   
15  63Ll5ofSkvOFp0BDwgHOm3  spotify:track:63Ll5ofSkvOFp0BDwgHOm3   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/0jIctjcQRcF1...   
1   https://api.spotify.com/v1/tracks/3PMcBzxNMaW4...   
2   https://api.spotify.com/v1/tracks/3LFzVLgW5N9R...   
3   https://api.spotify.com/v1/tracks/7xapl7fsqoJU...   
4   https://api.spotify.com/v1/tracks/2qqn52sp9dDk...   
5   https://api.spotify.com/v1/tracks/00SsZ9KAuNQB...  

### Album 5: Don't Panic: It's Longer Now

In [28]:
DontPanic_id = '7b3JowMC8FK0nToMlD4H1d'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(DontPanic_id)
DontPanic_tracks = requests.get(album_tracks_endpoint, headers = session_header)
DontPanic_dict = json.loads(DontPanic_tracks.text)
DontPanic_track_ids = [x['id'] for x in DontPanic_dict['items']]
DontPanic_track_ids
DontPanic_track_strings = ','.join(str(x) for x in DontPanic_track_ids)
DirtyWork_track_strings
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(DontPanic_track_strings)
DontPanic_features = requests.get(track_features_endpoint, headers = session_header)
DontPanic_dict = json.loads(DontPanic_features.text)

In [29]:
DontPanic_dict['audio_features'][0]

{'danceability': 0.459,
 'energy': 0.993,
 'key': 1,
 'loudness': -2.365,
 'mode': 1,
 'speechiness': 0.126,
 'acousticness': 0.000299,
 'instrumentalness': 0,
 'liveness': 0.214,
 'valence': 0.642,
 'tempo': 160.108,
 'type': 'audio_features',
 'id': '0aZ3B64WcYtvzvTz6GE7xT',
 'uri': 'spotify:track:0aZ3B64WcYtvzvTz6GE7xT',
 'track_href': 'https://api.spotify.com/v1/tracks/0aZ3B64WcYtvzvTz6GE7xT',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/0aZ3B64WcYtvzvTz6GE7xT',
 'duration_ms': 198925,
 'time_signature': 4}

In [30]:
DontPanic_df = pd.DataFrame(DontPanic_dict['audio_features'])

In [31]:
DontPanic_df['album'] = ['DontPanic'] * len(DontPanic_df.index)

In [32]:
DontPanic_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.459   0.993    1    -2.365     1       0.1260      0.000299   
1          0.475   0.987    5    -3.515     0       0.1210      0.000280   
2          0.511   0.964    6    -3.712     1       0.0451      0.000059   
3          0.436   0.977    2    -4.091     1       0.1050      0.000073   
4          0.423   0.980    7    -3.538     1       0.0717      0.000097   
5          0.572   0.983    8    -2.848     1       0.0828      0.000161   
6          0.405   0.979    4    -3.159     1       0.2130      0.000128   
7          0.572   0.982    2    -3.218     1       0.0765      0.000182   
8          0.427   0.986    6    -3.232     1       0.0857      0.002280   
9          0.406   0.987    6    -2.955     1       0.0629      0.000161   
10         0.534   0.989   11    -2.952     0       0.1170      0.001010   
11         0.498   0.981   11    -3.804     0       0.0680      0.000115   
12         0.369   0.966   11    -4.402     1       0.0695      0.000064   
13         0.415   0.980    1    -3.994     0       0.0838      0.000050   
14         0.428   0.983    1    -3.741     1       0.0890      0.000034   
15         0.524   0.895    7    -4.625     0       0.0423      0.008280   
16         0.534   0.476    3    -7.773     0       0.0278      0.101000   
17         0.610   0.475    1    -7.200     1       0.0279      0.130000   
18         0.553   0.454    1    -7.042     1       0.0279      0.131000   
19         0.641   0.458    5    -6.315     1       0.0272      0.211000   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000000    0.2140    0.642  160.108  audio_features   
1           0.000002    0.3350    0.632  162.005  audio_features   
2           0.000033    0.1720    0.546  140.050  audio_features   
3           0.000139    0.3090    0.586  165.902  audio_features   
4           0.000000    0.2880    0.593  175.154  audio_features   
5           0.000005    0.3120    0.552  127.970  audio_features   
6           0.000003    0.0743    0.334  190.098  audio_features   
7           0.000000    0.0629    0.642  132.032  audio_features   
8           0.000000    0.3110    0.619  178.224  audio_features   
9           0.000000    0.2160    0.662  175.010  audio_features   
10          0.000000    0.1290    0.493  129.972  audio_features   
11          0.000000    0.1550    0.454   92.988  audio_features   
12          0.000000    0.0571    0.465  166.782  audio_features   
13          0.000082    0.2860    0.569  160.082  audio_features   
14          0.000346    0.2240    0.639  174.998  audio_features   
15          0.000000    0.0879    0.557  157.052  audio_features   
16          0.000000    0.0488    0.506  150.050  audio_features   
17          0.000000    0.0802    0.436  123.020  audio_features   
18          0.000000    0.0981    0.440  140.023  audio_features   
19          0.000000    0.1070    0.494  130.026  audio_features   

                        id                                   uri  \
0   0aZ3B64WcYtvzvTz6GE7xT  spotify:track:0aZ3B64WcYtvzvTz6GE7xT   
1   0TPF6uW7fhMXJPoBERqTy5  spotify:track:0TPF6uW7fhMXJPoBERqTy5   
2   7dv7zzot4FkA9ovZ51yWyE  spotify:track:7dv7zzot4FkA9ovZ51yWyE   
3   1eMQ6C3nwItabEqSZocs4H  spotify:track:1eMQ6C3nwItabEqSZocs4H   
4   677GXVOkJpd6VEfsMJJuEM  spotify:track:677GXVOkJpd6VEfsMJJuEM   
5   1mYxMInBaaOcCqEaOc5FWg  spotify:track:1mYxMInBaaOcCqEaOc5FWg   
6   1DRNpWJUK7ycfhplV76HGq  spotify:track:1DRNpWJUK7ycfhplV76HGq   
7   6WM8fnjZKo7aq6LHWm0dOP  spotify:track:6WM8fnjZKo7aq6LHWm0dOP   
8   5hYu1CLzyRxL2jYV0UlhzM  spotify:track:5hYu1CLzyRxL2jYV0UlhzM   
9   1qi9S1UgCCRuZjFfIAYU4W  spotify:track:1qi9S1UgCCRuZjFfIAYU4W   
10  6wWEz11c2qDrqNbNcSlMtb  spotify:track:6wWEz11c2qDrqNbNcSlMtb   
11  6DceWz4d7szkZ6qxjS8TW0  spotify:track:6DceWz4d7szkZ6qxjS8TW0   
12  6POu6UT1J2PWXzl5d3YYEL  spotify:track:6POu6UT1J2PWXzl5d3YYEL   
13  7vORBjmxYvH5eWUJ4iYk6Q

### Album 6: Future Hearts

In [33]:
FutureHearts_id = '712X8QUI6uWg1JxHmz0l4I'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(FutureHearts_id)
FutureHearts_tracks = requests.get(album_tracks_endpoint, headers = session_header)
FutureHearts_dict = json.loads(FutureHearts_tracks.text)
FutureHearts_track_ids = [x['id'] for x in FutureHearts_dict['items']]
FutureHearts_track_ids
FutureHearts_track_strings = ','.join(str(x) for x in FutureHearts_track_ids)
FutureHearts_track_strings
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(FutureHearts_track_strings)
FutureHearts_features = requests.get(track_features_endpoint, headers = session_header)
FutureHearts_dict = json.loads(FutureHearts_features.text)

In [34]:
FutureHearts_dict['audio_features'][0]

{'danceability': 0.442,
 'energy': 0.842,
 'key': 2,
 'loudness': -6.101,
 'mode': 1,
 'speechiness': 0.0606,
 'acousticness': 0.0758,
 'instrumentalness': 1.84e-06,
 'liveness': 0.403,
 'valence': 0.318,
 'tempo': 72.558,
 'type': 'audio_features',
 'id': '5iN7ZEHuZIHMPxQ9Eb8RXN',
 'uri': 'spotify:track:5iN7ZEHuZIHMPxQ9Eb8RXN',
 'track_href': 'https://api.spotify.com/v1/tracks/5iN7ZEHuZIHMPxQ9Eb8RXN',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/5iN7ZEHuZIHMPxQ9Eb8RXN',
 'duration_ms': 144649,
 'time_signature': 4}

In [35]:
FutureHearts_df = pd.DataFrame(FutureHearts_dict['audio_features'])

In [36]:
FutureHearts_df['album'] = ['FutureHearts'] * len(FutureHearts_df.index)

In [37]:
FutureHearts_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.442   0.842    2    -6.101     1       0.0606      0.075800   
1          0.403   0.995    6    -3.007     1       0.1110      0.000097   
2          0.496   0.913    4    -3.286     1       0.0449      0.010100   
3          0.519   0.981    1    -3.472     0       0.0733      0.001370   
4          0.510   0.984    4    -4.126     1       0.1290      0.006040   
5          0.560   0.731    2    -6.121     1       0.0281      0.047200   
6          0.490   0.966    7    -3.677     1       0.1060      0.011100   
7          0.220   0.873    2    -4.364     1       0.0526      0.028500   
8          0.467   0.924   10    -4.294     1       0.2300      0.017400   
9          0.698   0.906    1    -3.921     1       0.0471      0.003780   
10         0.258   0.967    5    -4.076     0       0.1020      0.001330   
11         0.516   0.897    0    -4.917     1       0.0465      0.001530   
12         0.391   0.967    2    -4.244     0       0.1670      0.001030   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000002    0.4030    0.318   72.558  audio_features   
1           0.001050    0.3330    0.374  165.028  audio_features   
2           0.000000    0.1180    0.675  165.986  audio_features   
3           0.000164    0.0903    0.765  130.043  audio_features   
4           0.000000    0.0705    0.359  137.988  audio_features   
5           0.000000    0.2820    0.400   99.978  audio_features   
6           0.000008    0.2010    0.593  157.942  audio_features   
7           0.000000    0.1100    0.360   73.274  audio_features   
8           0.000000    0.0745    0.588  180.093  audio_features   
9           0.000027    0.2340    0.964  127.044  audio_features   
10          0.000002    0.1050    0.695  176.988  audio_features   
11          0.000000    0.1140    0.432  135.031  audio_features   
12          0.000000    0.0548    0.504  175.034  audio_features   

                        id                                   uri  \
0   5iN7ZEHuZIHMPxQ9Eb8RXN  spotify:track:5iN7ZEHuZIHMPxQ9Eb8RXN   
1   7trmqDnI68Vfa6qiFAumeT  spotify:track:7trmqDnI68Vfa6qiFAumeT   
2   1SeefzwSDiFCjRWaBslRIj  spotify:track:1SeefzwSDiFCjRWaBslRIj   
3   50oS45iNMFp7pKIQRAWIUP  spotify:track:50oS45iNMFp7pKIQRAWIUP   
4   60d8m5JYa6vHGg8hPFBZwQ  spotify:track:60d8m5JYa6vHGg8hPFBZwQ   
5   0LSY3NePenK0VG7nRUTVk8  spotify:track:0LSY3NePenK0VG7nRUTVk8   
6   4pVa7HzXbi4LZ8voROSZmy  spotify:track:4pVa7HzXbi4LZ8voROSZmy   
7   2CmZ0haTZB1Y1SCcFRu85Z  spotify:track:2CmZ0haTZB1Y1SCcFRu85Z   
8   52DDgK0wy9NnKMrw4A7MAJ  spotify:track:52DDgK0wy9NnKMrw4A7MAJ   
9   6CbDkmeFkd994Jx33xpg2Y  spotify:track:6CbDkmeFkd994Jx33xpg2Y   
10  3nf8hgFK0tMzvjGWkAnS9s  spotify:track:3nf8hgFK0tMzvjGWkAnS9s   
11  0f0YkUs2Vrd70LyG5031Wz  spotify:track:0f0YkUs2Vrd70LyG5031Wz   
12  5LU2hFmV7fQPmESUrMdf7n  spotify:track:5LU2hFmV7fQPmESUrMdf7n   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/5iN7ZEHuZIHM...   
1   https://api.spotify.com/v1/tracks/7trmqDnI68Vf...   
2   https://api.spotify.com/v1/tracks/1SeefzwSDiFC...   
3   https://api.spotify.com/v1/tracks/50oS45iNMFp7...   
4   https://api.spotify.com/v1/tracks/60d8m5JYa6vH...   
5   https://api.spotify.com/v1/tracks/0LSY3NePenK0...   
6   https://api.spotify.com/v1/tracks/4pVa7HzXbi4L...   
7   https://api.spotify.com/v1/tracks/2CmZ0haTZB1Y...   
8   https://api.spotify.com/v1/tracks/52DDgK0wy9Nn...   
9   https://api.spotify.com/v1/tracks/6CbDkmeFkd99...   
10  https://api.spotify.com/v1/tracks/3nf8hgFK0tMz...   
11  https://api.spotify.com/v1/tracks/0f0YkUs2Vrd7...   
12  https://api.spotify.com/v1/tracks/5LU2hFmV7fQP...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/5iN7...       144649   
1   https://api.spotify.com/v1/audio-analysis/7trm...       206084   
2   https://api.spotify.

### Album 7: Last Young Renegade

In [38]:
LastYoungRenegade_id = '0gpNGTVNivS2wB32tzV3OH'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(LastYoungRenegade_id)
LastYoungRenegade_tracks = requests.get(album_tracks_endpoint, headers = session_header)
LastYoungRenegade_dict = json.loads(LastYoungRenegade_tracks.text)
LastYoungRenegade_track_ids = [x['id'] for x in LastYoungRenegade_dict['items']]
LastYoungRenegade_track_ids
LastYoungRenegade_track_strings = ','.join(str(x) for x in LastYoungRenegade_track_ids)
LastYoungRenegade_track_strings
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(LastYoungRenegade_track_strings)
LastYoungRenegade_features = requests.get(track_features_endpoint, headers = session_header)
LastYoungRenegade_dict = json.loads(LastYoungRenegade_features.text)

In [39]:
LastYoungRenegade_dict['audio_features'][0]

{'danceability': 0.582,
 'energy': 0.843,
 'key': 2,
 'loudness': -5.596,
 'mode': 1,
 'speechiness': 0.0401,
 'acousticness': 0.00111,
 'instrumentalness': 0,
 'liveness': 0.102,
 'valence': 0.446,
 'tempo': 128.994,
 'type': 'audio_features',
 'id': '66hOCG7IpEJI5wfLDhajxD',
 'uri': 'spotify:track:66hOCG7IpEJI5wfLDhajxD',
 'track_href': 'https://api.spotify.com/v1/tracks/66hOCG7IpEJI5wfLDhajxD',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/66hOCG7IpEJI5wfLDhajxD',
 'duration_ms': 214093,
 'time_signature': 4}

In [40]:
LastYoungRenegade_df = pd.DataFrame(LastYoungRenegade_dict['audio_features'])

In [41]:
LastYoungRenegade_df['album'] = ['LastYoungRenegade'] * len(LastYoungRenegade_df.index)

In [42]:
LastYoungRenegade_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.582   0.843    2    -5.596     1       0.0401      0.001110   
1         0.490   0.868   10    -3.943     1       0.0375      0.048600   
2         0.560   0.751    4    -6.022     0       0.0364      0.001680   
3         0.560   0.887    7    -6.314     1       0.0594      0.003880   
4         0.462   0.952    5    -5.011     0       0.0529      0.000158   
5         0.553   0.830    6    -5.697     0       0.0380      0.000098   
6         0.537   0.624    0    -5.873     1       0.0288      0.075500   
7         0.515   0.780    5    -5.851     1       0.0478      0.000724   
8         0.640   0.782    9    -6.562     1       0.0352      0.004000   
9         0.580   0.629    2    -7.296     1       0.0391      0.029000   

   instrumentalness  liveness  valence    tempo            type  \
0          0.000000    0.1020    0.446  128.994  audio_features   
1          0.000000    0.0575    0.472  147.979  audio_features   
2          0.000052    0.1680    0.177   93.969  audio_features   
3          0.000000    0.3410    0.198  130.026  audio_features   
4          0.000192    0.0663    0.456  169.910  audio_features   
5          0.000000    0.3430    0.236  101.994  audio_features   
6          0.000000    0.1380    0.468  135.987  audio_features   
7          0.000007    0.0649    0.501  130.005  audio_features   
8          0.000033    0.5320    0.393  114.957  audio_features   
9          0.000002    0.0888    0.304   97.971  audio_features   

                       id                                   uri  \
0  66hOCG7IpEJI5wfLDhajxD  spotify:track:66hOCG7IpEJI5wfLDhajxD   
1  1isQMKYmpvJ4tOP8R2qkWr  spotify:track:1isQMKYmpvJ4tOP8R2qkWr   
2  7GZKme55JpmLq3alBDQm11  spotify:track:7GZKme55JpmLq3alBDQm11   
3  3aAoy02NwrmZcqXYYjqNh2  spotify:track:3aAoy02NwrmZcqXYYjqNh2   
4  4BpSuJ9GPFb4lNNNsOktws  spotify:track:4BpSuJ9GPFb4lNNNsOktws   
5  4pcyWJiAvOwdVxn7xjhR0N  spotify:track:4pcyWJiAvOwdVxn7xjhR0N   
6  0LJp1imLp25L6loDVPDMGn  spotify:track:0LJp1imLp25L6loDVPDMGn   
7  0ergYUxazsLy0qWK3h0YJd  spotify:track:0ergYUxazsLy0qWK3h0YJd   
8  2fF3XzFS37084kw9tgNA5i  spotify:track:2fF3XzFS37084kw9tgNA5i   
9  4nou22UEIrqNbtS9ZY3wjL  spotify:track:4nou22UEIrqNbtS9ZY3wjL   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/66hOCG7IpEJI...   
1  https://api.spotify.com/v1/tracks/1isQMKYmpvJ4...   
2  https://api.spotify.com/v1/tracks/7GZKme55JpmL...   
3  https://api.spotify.com/v1/tracks/3aAoy02NwrmZ...   
4  https://api.spotify.com/v1/tracks/4BpSuJ9GPFb4...   
5  https://api.spotify.com/v1/tracks/4pcyWJiAvOwd...   
6  https://api.spotify.com/v1/tracks/0LJp1imLp25L...   
7  https://api.spotify.com/v1/tracks/0ergYUxazsLy...   
8  https://api.spotify.com/v1/tracks/2fF3XzFS3708...   
9  https://api.spotify.com/v1/tracks/4nou22UEIrqN...   

                                        analysis_url  duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/66hO...       214093   
1  https://api.spotify.com/v1/audio-analysis/1isQ...       217680   
2  https://api.spotify.com/v1/audio-analysis/7GZK...       197640   
3  https://api.spotify.com/v1/audio-analysis/3aAo...       224947   
4  https://api.spotify.com/v1/audio-analysis/4BpS...       196507   
5  https://api.spotify.com/v1/audio-analysis/4pcy...       205307   
6  https://api.spotify.com/v1/audio-analysis/0LJp...       247773   
7  https://api.spotify.com/v1/audio-analysis/0erg...       206307   
8  https://api.spotify.com/v1/audio-analysis/2fF3...       236307   
9  https://api.spotify.com/v1/audio-analysis/4nou...       243560   

   time_signature              album  
0               4  LastYoungRenegade  
1               4  LastYoungRenegade  
2               4  LastYoungRenegade  
3               4  LastYoungRenegade  
4               4  LastYoungRenegade  
5               4  LastYoungRenegade  
6               4  LastYoungRenegade  
7           

### Album 8: Wake Up Sunshine

In [55]:
WakeUpSunshine_id = '1kwAv74rVTTGMpawGsXtiE'
album_tracks_endpoint = "https://api.spotify.com/v1/albums/{}/tracks".format(WakeUpSunshine_id)
WakeUpSunshine_tracks = requests.get(album_tracks_endpoint, headers = session_header)
WakeUpSunshine_dict = json.loads(WakeUpSunshine_tracks.text)
WakeUpSunshine_track_ids = [x['id'] for x in WakeUpSunshine_dict['items']]
WakeUpSunshine_track_ids
WakeUpSunshine_track_strings = ','.join(str(x) for x in WakeUpSunshine_track_ids)
WakeUpSunshine_track_strings  
track_features_endpoint = "https://api.spotify.com/v1/audio-features?ids={}".format(WakeUpSunshine_track_strings)
WakeUpSunshine_features = requests.get(track_features_endpoint, headers = session_header)
WakeUpSunshine_dict = json.loads(WakeUpSunshine_features.text)


In [59]:
WakeUpSunshine_dict['audio_features'][0]

{'danceability': 0.55,
 'energy': 0.837,
 'key': 11,
 'loudness': -5.411,
 'mode': 1,
 'speechiness': 0.044,
 'acousticness': 0.0104,
 'instrumentalness': 0,
 'liveness': 0.306,
 'valence': 0.637,
 'tempo': 89.821,
 'type': 'audio_features',
 'id': '02lU0Hj680zVexT4GqJJZ4',
 'uri': 'spotify:track:02lU0Hj680zVexT4GqJJZ4',
 'track_href': 'https://api.spotify.com/v1/tracks/02lU0Hj680zVexT4GqJJZ4',
 'analysis_url': 'https://api.spotify.com/v1/audio-analysis/02lU0Hj680zVexT4GqJJZ4',
 'duration_ms': 223994,
 'time_signature': 4}

In [48]:
WakeUpSunshine_df = pd.DataFrame(WakeUpSunshine_dict['audio_features'])

In [49]:
WakeUpSunshine_df['album'] = ['WakeUpSunshine'] * len(WakeUpSunshine_df.index)

In [50]:
WakeUpSunshine_df

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.550   0.837   11    -5.411     1       0.0440      0.010400   
1          0.656   0.955    7    -3.842     1       0.0462      0.008650   
2          0.556   0.899    4    -3.372     1       0.0711      0.002510   
3          0.643   0.927    2    -2.820     1       0.0533      0.001930   
4          0.511   0.903    6    -4.194     0       0.0685      0.000707   
5          0.508   0.951   11    -2.750     1       0.0709      0.001250   
6          0.369   0.845    0    -3.449     0       0.0611      0.049300   
7          0.686   0.591    3    -8.365     1       0.0450      0.569000   
8          0.598   0.886    8    -4.203     1       0.0412      0.000588   
9          0.583   0.859    1    -3.568     1       0.0432      0.002360   
10         0.473   0.884    1    -2.532     1       0.0841      0.008030   
11         0.563   0.907    1    -3.985     0       0.1260      0.001740   
12         0.543   0.699   10    -4.504     1       0.0349      0.078400   
13         0.561   0.943    2    -3.869     1       0.0689      0.004620   
14         0.628   0.819    4    -4.621     1       0.0309      0.122000   

    instrumentalness  liveness  valence    tempo            type  \
0           0.000000    0.3060    0.637   89.821  audio_features   
1           0.000000    0.3420    0.558  105.031  audio_features   
2           0.000000    0.0800    0.831  102.056  audio_features   
3           0.000000    0.1030    0.682  102.037  audio_features   
4           0.000000    0.0725    0.882  175.021  audio_features   
5           0.000000    0.2000    0.466  146.048  audio_features   
6           0.000000    0.4160    0.782   78.532  audio_features   
7           0.000813    0.3660    0.465  129.015  audio_features   
8           0.000048    0.0264    0.648  129.972  audio_features   
9           0.000000    0.1900    0.483  122.976  audio_features   
10          0.000000    0.3480    0.812  165.104  audio_features   
11          0.000000    0.4890    0.598  125.034  audio_features   
12          0.000000    0.0530    0.711   84.027  audio_features   
13          0.000000    0.2640    0.735  101.995  audio_features   
14          0.001310    0.3110    0.951   92.459  audio_features   

                        id                                   uri  \
0   02lU0Hj680zVexT4GqJJZ4  spotify:track:02lU0Hj680zVexT4GqJJZ4   
1   402Ecma69MRxeQK6ml6YbI  spotify:track:402Ecma69MRxeQK6ml6YbI   
2   6lMPUipRF49L8AxLW3F40B  spotify:track:6lMPUipRF49L8AxLW3F40B   
3   1wQsTZ4AxGoXB9FmD9Nf4o  spotify:track:1wQsTZ4AxGoXB9FmD9Nf4o   
4   4nOi9YbtFt7e9I4bbI1FKd  spotify:track:4nOi9YbtFt7e9I4bbI1FKd   
5   0xUvmXH7JO3klazeWlOyup  spotify:track:0xUvmXH7JO3klazeWlOyup   
6   0tyR7Bu9P086aWBFZ4QJoo  spotify:track:0tyR7Bu9P086aWBFZ4QJoo   
7   2nZ3LmQbY2rEd4awJVkQpd  spotify:track:2nZ3LmQbY2rEd4awJVkQpd   
8   2uVNn2pFpDlx1XrFcEfary  spotify:track:2uVNn2pFpDlx1XrFcEfary   
9   3IDgkcLETJkpp43cad5SwB  spotify:track:3IDgkcLETJkpp43cad5SwB   
10  2UQ7yB7jUsLTf3lNfrL2QV  spotify:track:2UQ7yB7jUsLTf3lNfrL2QV   
11  2z81IGdOL54S59AUdcFkw5  spotify:track:2z81IGdOL54S59AUdcFkw5   
12  4LpT15R6GqcKGjT92R8kfS  spotify:track:4LpT15R6GqcKGjT92R8kfS   
13  7AQG7amrUS84JzsarjS28y  spotify:track:7AQG7amrUS84JzsarjS28y   
14  3Aq9W9BBCjsFOQqcYyO6IA  spotify:track:3Aq9W9BBCjsFOQqcYyO6IA   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/02lU0Hj680zV...   
1   https://api.spotify.com/v1/tracks/402Ecma69MRx...   
2   https://api.spotify.com/v1/tracks/6lMPUipRF49L...   
3   https://api.spotify.com/v1/tracks/1wQsTZ4AxGoX...   
4   https://api.spotify.com/v1/tracks/4nOi9YbtFt7e...   
5   https://api.spotify.com/v1/tracks/0xUvmXH7JO3k...   
6   https://api.spotify.com/v1/tracks/0tyR7Bu9P086...   
7   https://api.spotify.com/v1/tracks/2nZ3LmQbY2rE...   
8   https://api.spotify.com/v1/tracks/2uVNn2pFpDlx...   
9   https://api.spotify.com/v1/tracks/3

In [51]:
audio_features = pd.concat([PutUp_df, SoWrongItsRight_df, NothingPersonal_df, DirtyWork_df, DontPanic_df, FutureHearts_df, LastYoungRenegade_df, WakeUpSunshine_df ])

In [52]:
audio_features

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.425   0.852    9    -6.148     1       0.1220      0.002280   
1          0.352   0.862    4    -6.521     1       0.0585      0.000042   
2          0.253   0.825    7    -6.903     1       0.0573      0.009480   
3          0.486   0.901    2    -6.943     1       0.0394      0.000049   
4          0.401   0.945    4    -5.984     1       0.2160      0.001880   
..           ...     ...  ...       ...   ...          ...           ...   
10         0.473   0.884    1    -2.532     1       0.0841      0.008030   
11         0.563   0.907    1    -3.985     0       0.1260      0.001740   
12         0.543   0.699   10    -4.504     1       0.0349      0.078400   
13         0.561   0.943    2    -3.869     1       0.0689      0.004620   
14         0.628   0.819    4    -4.621     1       0.0309      0.122000   

    instrumentalness  liveness  valence    tempo            type  \
0            0.00000    0.0963    0.431  185.021  audio_features   
1            0.00000    0.2980    0.491  179.769  audio_features   
2            0.00000    0.1140    0.115  172.679  audio_features   
3            0.00000    0.0828    0.353   96.030  audio_features   
4            0.00000    0.3250    0.408  174.953  audio_features   
..               ...       ...      ...      ...             ...   
10           0.00000    0.3480    0.812  165.104  audio_features   
11           0.00000    0.4890    0.598  125.034  audio_features   
12           0.00000    0.0530    0.711   84.027  audio_features   
13           0.00000    0.2640    0.735  101.995  audio_features   
14           0.00131    0.3110    0.951   92.459  audio_features   

                        id                                   uri  \
0   3j1HTRodSjGtK3YsYMI0XF  spotify:track:3j1HTRodSjGtK3YsYMI0XF   
1   6b23XfrLswilUCsHQHPY1i  spotify:track:6b23XfrLswilUCsHQHPY1i   
2   5ZEQLXv9F2pKtQrGwMKpMB  spotify:track:5ZEQLXv9F2pKtQrGwMKpMB   
3   1oTtOTbcP8waTv4wPsCqSm  spotify:track:1oTtOTbcP8waTv4wPsCqSm   
4   6f7h261u7CPBNFRBYL5Boq  spotify:track:6f7h261u7CPBNFRBYL5Boq   
..                     ...                                   ...   
10  2UQ7yB7jUsLTf3lNfrL2QV  spotify:track:2UQ7yB7jUsLTf3lNfrL2QV   
11  2z81IGdOL54S59AUdcFkw5  spotify:track:2z81IGdOL54S59AUdcFkw5   
12  4LpT15R6GqcKGjT92R8kfS  spotify:track:4LpT15R6GqcKGjT92R8kfS   
13  7AQG7amrUS84JzsarjS28y  spotify:track:7AQG7amrUS84JzsarjS28y   
14  3Aq9W9BBCjsFOQqcYyO6IA  spotify:track:3Aq9W9BBCjsFOQqcYyO6IA   

                                           track_href  \
0   https://api.spotify.com/v1/tracks/3j1HTRodSjGt...   
1   https://api.spotify.com/v1/tracks/6b23XfrLswil...   
2   https://api.spotify.com/v1/tracks/5ZEQLXv9F2pK...   
3   https://api.spotify.com/v1/tracks/1oTtOTbcP8wa...   
4   https://api.spotify.com/v1/tracks/6f7h261u7CPB...   
..                                                ...   
10  https://api.spotify.com/v1/tracks/2UQ7yB7jUsLT...   
11  https://api.spotify.com/v1/tracks/2z81IGdOL54S...   
12  https://api.spotify.com/v1/tracks/4LpT15R6GqcK...   
13  https://api.spotify.com/v1/tracks/7AQG7amrUS84...   
14  https://api.spotify.com/v1/tracks/3Aq9W9BBCjsF...   

                                         analysis_url  duration_ms  \
0   https://api.spotify.com/v1/audio-analysis/3j1H...       181027   
1   https://api.spotify.com/v1/audio-analysis/6b23...       183573   
2   https://api.spotify.com/v1/audio-analysis/5ZEQ...       218200   
3   https://api.spotify.com/v1/audio-analysis/1oTt...       218293   
4   https://api.spotify.com/v1/audio-analysis/6f7h...       177840   
..                                                ...          ...   
10  https://api.spotify.com/v1/audio-analysis/2UQ7...       166454   
11  https://api.spotify.com/v1/audio-analysis/2z81...       182067   
12  https://api.spotify.com/v1/audio-analysis/4LpT...       183167   
13  https://api.spotify.com/v1/audio-analysis/7AQG...       194990   
14

In [53]:
audio_features.to_csv('audio_features.csv')

In [54]:
audio_features.sort_values('danceability', ascending = False).head(10)

danceability  energy  key  loudness  mode  speechiness  acousticness  \
9          0.698   0.906    1    -3.921     1       0.0471      0.003780   
7          0.686   0.591    3    -8.365     1       0.0450      0.569000   
1          0.656   0.955    7    -3.842     1       0.0462      0.008650   
3          0.643   0.927    2    -2.820     1       0.0533      0.001930   
19         0.641   0.458    5    -6.315     1       0.0272      0.211000   
8          0.640   0.782    9    -6.562     1       0.0352      0.004000   
14         0.628   0.819    4    -4.621     1       0.0309      0.122000   
2          0.624   0.928    4    -3.022     1       0.0575      0.011300   
17         0.610   0.475    1    -7.200     1       0.0279      0.130000   
8          0.598   0.886    8    -4.203     1       0.0412      0.000588   

    instrumentalness  liveness  valence    tempo            type  \
9           0.000027    0.2340    0.964  127.044  audio_features   
7           0.000813    0.3660    0.465  129.015  audio_features   
1           0.000000    0.3420    0.558  105.031  audio_features   
3           0.000000    0.1030    0.682  102.037  audio_features   
19          0.000000    0.1070    0.494  130.026  audio_features   
8           0.000033    0.5320    0.393  114.957  audio_features   
14          0.001310    0.3110    0.951   92.459  audio_features   
2           0.000000    0.1020    0.671  132.045  audio_features   
17          0.000000    0.0802    0.436  123.020  audio_features   
8           0.000048    0.0264    0.648  129.972  audio_features   

                        id                                   uri  \
9   6CbDkmeFkd994Jx33xpg2Y  spotify:track:6CbDkmeFkd994Jx33xpg2Y   
7   2nZ3LmQbY2rEd4awJVkQpd  spotify:track:2nZ3LmQbY2rEd4awJVkQpd   
1   402Ecma69MRxeQK6ml6YbI  spotify:track:402Ecma69MRxeQK6ml6YbI   
3   1wQsTZ4AxGoXB9FmD9Nf4o  spotify:track:1wQsTZ4AxGoXB9FmD9Nf4o   
19  1cyp7q6L1RMbACJVr01DWU  spotify:track:1cyp7q6L1RMbACJVr01DWU   
8   2fF3XzFS37084kw9tgNA5i  spotify:track:2fF3XzFS37084kw9tgNA5i   
14  3Aq9W9BBCjsFOQqcYyO6IA  spotify:track:3Aq9W9BBCjsFOQqcYyO6IA   
2   3LFzVLgW5N9RYirdQOOPX8  spotify:track:3LFzVLgW5N9RYirdQOOPX8   
17  5CVXM0oPrmzZykiMP6FMr5  spotify:track:5CVXM0oPrmzZykiMP6FMr5   
8   2uVNn2pFpDlx1XrFcEfary  spotify:track:2uVNn2pFpDlx1XrFcEfary   

                                           track_href  \
9   https://api.spotify.com/v1/tracks/6CbDkmeFkd99...   
7   https://api.spotify.com/v1/tracks/2nZ3LmQbY2rE...   
1   https://api.spotify.com/v1/tracks/402Ecma69MRx...   
3   https://api.spotify.com/v1/tracks/1wQsTZ4AxGoX...   
19  https://api.spotify.com/v1/tracks/1cyp7q6L1RMb...   
8   https://api.spotify.com/v1/tracks/2fF3XzFS3708...   
14  https://api.spotify.com/v1/tracks/3Aq9W9BBCjsF...   
2   https://api.spotify.com/v1/tracks/3LFzVLgW5N9R...   
17  https://api.spotify.com/v1/tracks/5CVXM0oPrmzZ...   
8   https://api.spotify.com/v1/tracks/2uVNn2pFpDlx...   

                                         analysis_url  duration_ms  \
9   https://api.spotify.com/v1/audio-analysis/6CbD...       212951   
7   https://api.spotify.com/v1/audio-analysis/2nZ3...       182451   
1   https://api.spotify.com/v1/audio-analysis/402E...       180599   
3   https://api.spotify.com/v1/audio-analysis/1wQs...       174895   
19  https://api.spotify.com/v1/audio-analysis/1cyp...       194811   
8   https://api.spotify.com/v1/audio-analysis/2fF3...       236307   
14  https://api.spotify.com/v1/audio-analysis/3Aq9...       185988   
2   https://api.spotify.com/v1/audio-analysis/3LFz...       169227   
17  https://api.spotify.com/v1/audio-analysis/5CVX...       220598   
8   https://api.spotify.com/v1/audio-analysis/2uVN...       193949   

    time_signature              album  
9                4       FutureHearts  
7                4     WakeUpSunshine  
1                4     WakeUpSunshine  
3                4     WakeUpSunshine  
19               4          DontPanic  
8                4  LastYoungR

## From the data above, I can conclude that of the top 10 most danceable songs, the album with the most songs from the top 10 is Wake Up Sunshine. This did not surprise me as their songs are very catchy and has some of the highest tempos of the top 10. 

### However, the most danceable song is not from Wake Up Sunshine, it is from the Album Future Hearts. After going back to Spotify and viewing the album and tracks as a whole, I counted the song number minus 1 since the first track starts at 0 and found that the name of the track is "Bail Me Out (feat Joel Madden)."

#### I am hoping to be able to display this evidence via code of the track name.

### I found another interesting fact, Bail Me Out is not the most listened to song on the album of Future Hearts, it is "Missing You." Bail Me out has 9,505,241 plays as of 12/8/2022 and Missing You has 68,449,902 plays. I would have thought that the most danceable song would be the most played of the album so this was an interesting find. 

#### I am hoping to be able to display this evidence via code of the number of plays.

#### I am also hoping to be able to compare with the popularity of the albums and tracks and compare them with danceability. 

#### I am hoping to be able to display the audio features as a whole for an album.

In [60]:
audio_features.sort_values('tempo', ascending = False).head(10)

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.312   0.953    4    -2.051     1       0.1300      0.003500   
6         0.405   0.979    4    -3.159     1       0.2130      0.000128   
0         0.425   0.852    9    -6.148     1       0.1220      0.002280   
1         0.512   0.947    1    -2.730     0       0.0817      0.006330   
6         0.318   0.936    2    -2.522     1       0.0489      0.001710   
1         0.373   0.952    9    -2.525     1       0.0813      0.002010   
0         0.497   0.923    2    -4.512     1       0.1380      0.002820   
5         0.459   0.895    2    -3.126     1       0.0805      0.007250   
8         0.467   0.924   10    -4.294     1       0.2300      0.017400   
3         0.411   0.970   11    -2.860     0       0.1070      0.004550   

   instrumentalness  liveness  valence    tempo            type  \
0          0.000000    0.3480    0.644  198.898  audio_features   
6          0.000003    0.0743    0.334  190.098  audio_features   
0          0.000000    0.0963    0.431  185.021  audio_features   
1          0.000000    0.0613    0.553  184.005  audio_features   
6          0.000000    0.0852    0.587  182.996  audio_features   
1          0.000000    0.0355    0.592  182.963  audio_features   
0          0.000000    0.1580    0.663  181.074  audio_features   
5          0.000000    0.2060    0.572  181.040  audio_features   
8          0.000000    0.0745    0.588  180.093  audio_features   
3          0.000000    0.1290    0.488  180.023  audio_features   

                       id                                   uri  \
0  2rQSAtxX1yuoWq8ZrvUhpL  spotify:track:2rQSAtxX1yuoWq8ZrvUhpL   
6  1DRNpWJUK7ycfhplV76HGq  spotify:track:1DRNpWJUK7ycfhplV76HGq   
0  3j1HTRodSjGtK3YsYMI0XF  spotify:track:3j1HTRodSjGtK3YsYMI0XF   
1  3PMcBzxNMaW4cVX7FKXwwA  spotify:track:3PMcBzxNMaW4cVX7FKXwwA   
6  5pWqD58cZ9TNvL9WbNfcQ3  spotify:track:5pWqD58cZ9TNvL9WbNfcQ3   
1  4NpzB4z4B0UcajopWE5e6h  spotify:track:4NpzB4z4B0UcajopWE5e6h   
0  0aZJGkkXR3DgaFqo5sB8ot  spotify:track:0aZJGkkXR3DgaFqo5sB8ot   
5  0JJP0IS4w0fJx01EcrfkDe  spotify:track:0JJP0IS4w0fJx01EcrfkDe   
8  52DDgK0wy9NnKMrw4A7MAJ  spotify:track:52DDgK0wy9NnKMrw4A7MAJ   
3  7xapl7fsqoJU4p0mQVOSm0  spotify:track:7xapl7fsqoJU4p0mQVOSm0   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/2rQSAtxX1yuo...   
6  https://api.spotify.com/v1/tracks/1DRNpWJUK7yc...   
0  https://api.spotify.com/v1/tracks/3j1HTRodSjGt...   
1  https://api.spotify.com/v1/tracks/3PMcBzxNMaW4...   
6  https://api.spotify.com/v1/tracks/5pWqD58cZ9TN...   
1  https://api.spotify.com/v1/tracks/4NpzB4z4B0Uc...   
0  https://api.spotify.com/v1/tracks/0aZJGkkXR3Dg...   
5  https://api.spotify.com/v1/tracks/0JJP0IS4w0fJ...   
8  https://api.spotify.com/v1/tracks/52DDgK0wy9Nn...   
3  https://api.spotify.com/v1/tracks/7xapl7fsqoJU...   

                                        analysis_url  duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/2rQS...       149613   
6  https://api.spotify.com/v1/audio-analysis/1DRN...       169015   
0  https://api.spotify.com/v1/audio-analysis/3j1H...       181027   
1  https://api.spotify.com/v1/audio-analysis/3PMc...       180600   
6  https://api.spotify.com/v1/audio-analysis/5pWq...       221733   
1  https://api.spotify.com/v1/audio-analysis/4Npz...       180440   
0  https://api.spotify.com/v1/audio-analysis/0aZJ...       198000   
5  https://api.spotify.com/v1/audio-analysis/0JJP...       182827   
8  https://api.spotify.com/v1/audio-analysis/52DD...       185735   
3  https://api.spotify.com/v1/audio-analysis/7xap...       197053   

   time_signature            album  
0               4  SoWrongItsRight  
6               4        DontPanic  
0               4            PutUp  
1               4        DirtyWork  
6               4  SoWrongItsRight  
1               4  SoWrongItsRight  
0               4  NothingPersonal  
5               4  SoWrongIt

#### It's interesting to compare that the songs with the highest tempo, have a pretty low danceability. It was especially surprising to see that the highest tempo of an All Time Low song is 198.898. That's insane! I found a metronome online and tested the tempo while listening to the song and it was crazy how fast it was! I learned about myself that I like songs with high tempo. 